File: src/main/java/wraith/fabricaeexnihilo/compatibility/rei/witchwater/WitchWaterEntityDisplay.java
package wraith.fabricaeexnihilo.compatibility.rei.witchwater;

import me.shedaniel.rei.api.common.category.CategoryIdentifier;
import me.shedaniel.rei.api.common.display.Display;
import me.shedaniel.rei.api.common.entry.EntryIngredient;
import net.minecraft.command.argument.NbtPathArgumentType;
import net.minecraft.entity.EntityType;
import net.minecraft.util.Identifier;
import wraith.fabricaeexnihilo.compatibility.rei.PluginEntry;
import wraith.fabricaeexnihilo.recipe.witchwater.WitchWaterEntityRecipe;

import java.util.List;
import java.util.Optional;

public class WitchWaterEntityDisplay implements Display {
    public final NbtPathArgumentType.NbtPath nbt;
    public final EntityType<?> result;
    public final EntityType<?> target;
    private final Identifier id;

    public WitchWaterEntityDisplay(WitchWaterEntityRecipe recipe) {
        this.target = recipe.getTarget();
        this.result = recipe.getResult();
        this.nbt = recipe.getNbt();
        this.id = recipe.getId();
    }

    @Override
    public CategoryIdentifier<?> getCategoryIdentifier() {
        return PluginEntry.WITCH_WATER_ENTITY;
    }

    @Override
    public List<EntryIngredient> getInputEntries() {
        return List.of();
    }

    @Override
    public List<EntryIngredient> getOutputEntries() {
        return List.of();
    }

    @Override
    public Optional<Identifier> getDisplayLocation() {
        return Optional.of(id);
    }
}

File: src/main/kotlin/exnihilofabrico/compatibility/rei/witchwater/WitchWaterEntityDisplay.kt
package exnihilofabrico.compatibility.rei.witchwater

import exnihilofabrico.api.recipes.witchwater.WitchWaterEntityRecipe
import exnihilofabrico.compatibility.rei.PluginEntry
import exnihilofabrico.modules.witchwater.WitchWaterFluid
import exnihilofabrico.util.asREIEntry
import me.shedaniel.rei.api.EntryStack
import me.shedaniel.rei.api.RecipeDisplay
import net.minecraft.item.SpawnEggItem

class WitchWaterEntityDisplay(val recipe: WitchWaterEntityRecipe): RecipeDisplay {
    override fun getRecipeCategory() = PluginEntry.WITCH_WATER_ENTITY

    override fun getOutputEntries() = listOf(SpawnEggItem.forEntity(recipe.tospawn)?.asREIEntry() ?: EntryStack.empty())
    override fun getInputEntries() =
        listOf(recipe.target.asREIEntries(), listOf(WitchWaterFluid.bucket.asREIEntry()))
}